src/lib/state/DY.Lib.State.Map.fst

```diff
@@ -15,75 +15,88 @@ open DY.Lib.State.Typed
 
-noeq type map_predicate {|crypto_invariants|} (mt:map_types) = {
-  pred: trace -> principal -> nat -> mt.key -> mt.value -> prop;
-  pred_later: tr1:trace -> tr2:trace -> prin:principal -> sess_id:nat -> key:mt.key -> value:mt.value -> Lemma
+noeq type map_predicate {|crypto_invariants|} (key_t:eqtype) (value_t:Type0) {|mt:map_types key_t value_t|} = {
+  pred: trace -> principal -> nat -> key_t -> value_t -> prop;
+  pred_later: tr1:trace -> tr2:trace -> prin:principal -> sess_id:nat -> key:key_t -> value:value_t -> Lemma
     (requires pred tr1 prin sess_id key value /\ tr1 <$ tr2)
     (ensures pred tr2 prin sess_id key value)
   ;
-  pred_knowable: tr:trace -> prin:principal -> sess_id:nat -> key:mt.key -> value:mt.value -> Lemma
+  pred_knowable: tr:trace -> prin:principal -> sess_id:nat -> key:key_t -> value:value_t -> Lemma
     (requires pred tr prin sess_id key value)
-    (ensures is_well_formed_prefix mt.ps_key (is_knowable_by (principal_state_label prin sess_id) tr) key /\ is_well_formed_prefix mt.ps_value (is_knowable_by (principal_state_label prin sess_id) tr) value)
+    (ensures is_well_formed_prefix mt.ps_key_t (is_knowable_by (principal_state_label prin sess_id) tr) key /\ is_well_formed_prefix mt.ps_value_t (is_knowable_by (principal_state_label prin sess_id) tr) value)
   ;
 }
 
 [@@ with_bytes bytes]
-noeq type map_elem (mt:map_types) = {
-  [@@@ with_parser #bytes mt.ps_key]
-  key: mt.key;
-  [@@@ with_parser #bytes mt.ps_value]
-  value: mt.value;
+noeq type map_elem (key_t:eqtype) (value_t:Type0) {|mt:map_types key_t value_t|} = {
+  [@@@ with_parser #bytes mt.ps_key_t]
+  key: key_t;
+  [@@@ with_parser #bytes mt.ps_value_t]
+  value: value_t;
 }
 
 %splice [ps_map_elem] (gen_parser (`map_elem))
 %splice [ps_map_elem_is_well_formed] (gen_is_well_formed_lemma (`map_elem))
 
 [@@ with_bytes bytes]
-noeq type map (mt:map_types) = {
-  [@@@ with_parser #bytes (ps_list (ps_map_elem mt))]
-  key_values: list (map_elem mt)
+noeq type map (key_t:eqtype) (value_t:Type0) {|mt:map_types key_t value_t|} = {
+  [@@@ with_parser #bytes (ps_list (ps_map_elem key_t value_t))]
+  key_values: list (map_elem key_t value_t)
 }
 
 %splice [ps_map] (gen_parser (`map))
 %splice [ps_map_is_well_formed] (gen_is_well_formed_lemma (`map))
 
-instance parseable_serializeable_map (mt:map_types) : parseable_serializeable bytes (map mt) = mk_parseable_serializeable (ps_map mt)
+instance parseable_serializeable_map (key_t:eqtype) (value_t:Type0) {|map_types key_t value_t|} : parseable_serializeable bytes (map key_t value_t) = mk_parseable_serializeable (ps_map key_t value_t)
 
-val map_elem_invariant: {|crypto_invariants|} -> #mt:map_types -> map_predicate mt -> trace -> principal -> nat -> map_elem mt -> prop
-let map_elem_invariant #cinvs #mt mpred tr prin sess_id x =
+instance map_has_local_state (key_t:eqtype) (value_t:Type0) {|mt:map_types key_t value_t|}: local_state (map key_t value_t) = {
```

Comment:
Again, the name of this instance sounds strange, in particular the `has`. To me `map_is_local_state` sounds more natural, however following your previously mentioned naming convention, I would expected something like `local_state_map`. (Maybe we should write down this naming convention somewhere.)

Reply:
I realize that I did not follow this convention everywhere, sometimes it is `typeclass_typename` and sometimes it is `typename_typeclass`… I will open an issue for that to remember it.

Reply:
If we have decided on one option, we should also add the naming convention to the CONTRIBUTING.md file.

Reply:
I agree, I opened the issue #20 to decide on that later, because the scope is larger than this pull-request.